src/UI/Buyer/src/app/shared/validators/date-input/date-input.validator.ts

```diff
@@ -0,0 +1,24 @@
+import { ValidatorFn, AbstractControl, ValidationErrors } from '@angular/forms';
+
+/**
+ * Our date inputs use ngbDatepicker but also allow freeform entry.
+ * We need to validate the free form entry strings which are converted to date objects
+ */
+
+export function DateValidator(): ValidatorFn {
+  return (control: AbstractControl): ValidationErrors | null => {
+    // only validate if both fields have been touched
+    if (control.value == null || control.value === '') {
+      return null;
+    }
+
+    if (
+      control.value instanceof Date &&
+      control.value.getFullYear().toString().length === 4
+    ) {
```

Comment:
do we want to validate length of month? lol not even sure if its applicable but thought itd be worth bringing up

Reply:
This is kinda a weird situation, I'll add more comments to the code. The string you enter is converted to a date object, so if you have a month with three numbers, or a date past 31 it will fail to convert to a date and this function will notice an error. However, there's no limit on years, so if you enter 999 or 10000, it will successfully convert to a date. I'm checking those because I think its safe to assume those are typos.